wallet/createtx.go

```diff
@@ -1654,14 +1659,42 @@ func (w *Wallet) purchaseTickets(ctx context.Context, op errors.Op,
 		// request first, then check the ticket address
 		// stored from the configuation. Finally, generate
 		// an address.
-		addrVote := req.VotingAddress
-		if addrVote == nil && req.CSPPServer == "" {
-			addrVote = w.ticketAddress
-		}
-		if addrVote == nil {
-			addrVote, err = stakeAddrFunc(op, req.VotingAccount, 1)
-			if err != nil {
-				return nil, err
+		var addrVote stdaddr.StakeAddress
+
+		// If req.UseVotingAccount is true, always take the submission
+		// script's address from the voting account. This is intended
+		// to be used with a special account type. The signing address
+		// for the same index is saved to the database. That address is
+		// later used to sign messages sent to a vspd related to this
+		// ticket.
+		if req.UseVotingAccount {
+			var idx uint32
+			for i := 0; ; i++ {
+				addrVote, idx, err = stakeAddrFunc(op, req.VotingAccount, 1)
+				if err != nil {
+					return nil, err
+				}
+				_, err := w.signingAddressAtIdx(ctx, op, w.persistReturnedChild(ctx, nil), req.VotingAccount, idx)
```

Comment:
You're calling `signingAddressAtIdx` with the same `idx` in every iteration of this loop, so this doesn't really try to generate a different key.

I haven't looked at the code that handles using the signing key yet, but trying to specially handle the case where this returns `hdkeychain.ErrInvalidChild` seems problematic in the sense that the code that handles the vsp signing ops will now have to also track whether it's that index or a different one (though I guess that particular problem might be abstracted away by the key persistence).

Also, this could lead to address reuse for the vsp checking key if you use two consecutive addresses for voting keys, where the first corresponding signing key returns `ErrInvalidChild`.

I think the odds of generating invalid keys are low enough that we can punt on trying to specially handle this case, simply return an error and let the user try a next attempt at purchasing the ticket, which will return the next index of the voting account and thus a different signing address.

Reply:
Will remove the loop.

Reply:
Removed.